File: packages/totem-sdk/packages/kissvm/src/mast/proof-chain.ts
/**
 * Proof chain builder — constructs and verifies multi-level recursive MAST
 * proof chains. Each link proves that a script is authorized by a policy
 * root, and the script may delegate to the next policy root.
 *
 * Proof verification delegates to the canonical MMR verifier in
 * mast-compiler.ts. The sorted-pair Merkle hashing previously used here
 * has been removed — all verification now uses Minima-compatible MMR proofs.
 */

import { sha3_256, bytesToHex } from '@totemsdk/core';
import { MiniNumber } from '../MiniNumber.js';
import { verifyScriptMembership } from './mast-compiler.js';
import type { ProofLink, ProofChain, VerificationResult } from './types.js';
export type { ProofLink, ProofChain, VerificationResult };

function hashScript(script: string): string {
  return bytesToHex(sha3_256(new TextEncoder().encode(script)));
}

export function buildProofChain(links: ProofLink[]): ProofChain {
  if (links.length === 0) throw new Error('Proof chain must have at least one link');

  for (const link of links) {
    const computedHash = hashScript(link.script);
    if (computedHash !== link.scriptHash) {
      throw new Error(`Script hash mismatch for "${link.label ?? 'unnamed'}": expected ${link.scriptHash}, got ${computedHash}`);
    }
  }

  return {
    links: [...links],
    depth: links.length,
    verified: false,
    leafScriptHash: links[links.length - 1].scriptHash,
  };
}

export function verifyProofChain(
  chain: ProofChain,
  expectedLeafScriptHash?: string,
): VerificationResult {
  if (chain.links.length === 0) {
    return { valid: false, reason: 'Empty proof chain' };
  }

  for (let i = 0; i < chain.links.length; i++) {
    const link = chain.links[i];

    const result = verifyScriptMembership(link.script, link.proof, link.policyRoot);
    if (!result.valid) {
      return {
        valid: false,
        failedAt: i,
        reason: `MMR proof verification failed at level ${i} ("${link.label ?? 'unnamed'}"): ${result.reason}`,
      };
    }

    if (i < chain.links.length - 1) {
      const nextLink = chain.links[i + 1];
      const mastRef = `MAST 0x${nextLink.policyRoot}`;
      if (!link.script.includes(mastRef) && !link.script.includes(`MAST ${nextLink.policyRoot}`)) {
        return {
          valid: false,
          failedAt: i,
          reason: `Delegation verification failed at level ${i} ("${link.label ?? 'unnamed'}"): script does not contain MAST referencing next root ${nextLink.policyRoot.slice(0, 16)}…`,
        };
      }
    }
  }

  if (expectedLeafScriptHash) {
    const leaf = chain.links[chain.links.length - 1];
    if (leaf.scriptHash !== expectedLeafScriptHash) {
      return {
        valid: false,
        failedAt: chain.links.length - 1,
        reason: `Leaf script hash mismatch: expected ${expectedLeafScriptHash.slice(0, 16)}…, got ${leaf.scriptHash.slice(0, 16)}…`,
      };
    }
  }

  chain.verified = true;
  return { valid: true, chain };
}

/**
 * Generate a canonical Minima 5-argument PROOF expression.
 *
 * Canonical Minima syntax: PROOF(data, leafSum, rootHash, rootSum, proofHex)
 *
 * @returns Minima expression: `PROOF(0x<scriptHash> <leafSum> 0x<policyRoot> <rootSum> 0x<proof>)`
 */
export function toMinimaProofExpression(link: ProofLink): string {
  const leafSum = link.leafSum ?? MiniNumber.ZERO;
  const rootSum = link.rootSum ?? MiniNumber.ZERO;
  return `PROOF(0x${link.scriptHash} ${leafSum} 0x${link.policyRoot} ${rootSum} 0x${link.proof})`;
}

/**
 * @deprecated Use toMinimaProofExpression(). Canonical Minima PROOF takes
 * five arguments: data, leafSum, rootHash, rootSum, proofHex.
 */
export function toTotemProofExpression(link: ProofLink): string {
  return toMinimaProofExpression(link);
}

/**
 * @deprecated Use toMinimaProofExpression(). Canonical Minima PROOF takes
 * five arguments: data, leafSum, rootHash, rootSum, proofHex.
 */
export function toProofExpression(link: ProofLink): string {
  return toMinimaProofExpression(link);
}

/**
 * Generate the full nested MAST KISSVM script for a proof chain.
 *
 * Each level uses `MAST 0x<root>` to auto-load the next script from the
 * transaction witness. The VM looks up the witness ScriptProof whose
 * calculated address equals the given root, parses it, and executes it
 * in the same contract context.
 *
 * VM limits: 64 stack depth, 1,024 instructions shared across all frames.
 *
 * @returns KISSVM script with nested MAST expressions.
 */
export function toNestedMastScript(chain: ProofChain): string {
  if (chain.links.length === 0) return 'RETURN TRUE';

  let script = chain.links[chain.links.length - 1].script;

  for (let i = chain.links.length - 2; i >= 0; i--) {
    const nextRoot = chain.links[i + 1].policyRoot;
    script = `${chain.links[i].script}\nMAST 0x${nextRoot}`;
  }

  return script;
}

File: packages/totem-sdk/packages/recursive-mast/rust/src/proof_chain.rs
use crate::content_keys::compute_script_hash;
use crate::types::{ProofChain, ProofLink, VerificationResult};

pub fn build_proof_chain(links: &[ProofLink]) -> Result<ProofChain, String> {
    if links.is_empty() {
        return Err("Proof chain must have at least one link".to_string());
    }

    for link in links {
        let computed_hash = compute_script_hash(&link.script);
        if computed_hash != link.script_hash {
            return Err(format!(
                "Script hash mismatch for \"{}\": expected {}, got {}",
                link.label.as_deref().unwrap_or("unnamed"),
                link.script_hash,
                computed_hash
            ));
        }
    }

    Ok(ProofChain {
        links: links.to_vec(),
        depth: links.len() as u32,
        verified: false,
        leaf_script_hash: links[links.len() - 1].script_hash.clone(),
    })
}

pub fn verify_proof_chain(
    chain: &ProofChain,
    expected_leaf_script_hash: Option<&str>,
    verify_membership_fn: &dyn Fn(&str, &str, &str) -> bool,
) -> VerificationResult {
    if chain.links.is_empty() {
        return VerificationResult {
            valid: false,
            failed_at: None,
            reason: Some("Empty proof chain".to_string()),
        };
    }

    for i in 0..chain.links.len() {
        let link = &chain.links[i];

        if !verify_membership_fn(&link.script, &link.proof, &link.policy_root) {
            return VerificationResult {
                valid: false,
                failed_at: Some(i as u32),
                reason: Some(format!(
                    "MMR proof verification failed at level {} (\"{}\")",
                    i,
                    link.label.as_deref().unwrap_or("unnamed")
                )),
            };
        }

        if i < chain.links.len() - 1 {
            let next_link = &chain.links[i + 1];
            let mast_ref = format!("MAST 0x{}", next_link.policy_root);
            if !link.script.contains(&mast_ref) && !link.script.contains(&format!("MAST {}", next_link.policy_root)) {
                return VerificationResult {
                    valid: false,
                    failed_at: Some(i as u32),
                    reason: Some(format!(
                        "Delegation verification failed at level {} (\"{}\"): script does not contain MAST referencing next root {}…",
                        i,
                        link.label.as_deref().unwrap_or("unnamed"),
                        &next_link.policy_root[..16.min(next_link.policy_root.len())]
                    )),
                };
            }
        }
    }

    if let Some(expected) = expected_leaf_script_hash {
        let leaf = &chain.links[chain.links.len() - 1];
        if leaf.script_hash != expected {
            return VerificationResult {
                valid: false,
                failed_at: Some((chain.links.len() - 1) as u32),
                reason: Some(format!(
                    "Leaf script hash mismatch: expected {}…, got {}…",
                    &expected[..16.min(expected.len())],
                    &leaf.script_hash[..16.min(leaf.script_hash.len())]
                )),
            };
        }
    }

    VerificationResult {
        valid: true,
        failed_at: None,
        reason: None,
    }
}

pub fn to_minima_proof_expression(link: &ProofLink) -> String {
    let leaf_sum = link.leaf_sum.as_deref().unwrap_or("0");
    let root_sum = link.root_sum.as_deref().unwrap_or("0");
    format!(
        "PROOF(0x{} {} 0x{} {} 0x{})",
        link.script_hash, leaf_sum, link.policy_root, root_sum, link.proof
    )
}

pub fn to_nested_mast_script(chain: &ProofChain) -> String {
    if chain.links.is_empty() {
        return "RETURN TRUE".to_string();
    }

    let mut script = chain.links[chain.links.len() - 1].script.clone();

    for i in (0..chain.links.len() - 1).rev() {
        let next_root = &chain.links[i + 1].policy_root;
        script = format!("{}\nMAST 0x{}", chain.links[i].script, next_root);
    }

    script
}
